### ISSUE TITLE
Dokany cannot survive basic Windows Sleep

### ISSUE THREAD

**Issue by KyleSanderson** [dokan-dev/dokany#634]
Feature request can skip this form. Bug report **must** complete it. `Check List` must be 100% match or it will be automatically closed without further discussion. Please remove this line.

# Environment
 * Windows version: Windows 10 2016 LTSB
 * Processor architecture: amd64
 * Dokany version: 
 * Library type (Dokany/FUSE): Dokany 10.0.5.1000 (Latest).
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description below.
  - [x] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [x] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

# Description
Dokany hard locks explorer.exe and becomes completely non-responsive after Windows goes to sleep and is awoken. 100% reproducible with WinSshFS Foreveryone. Dokany even hard locks on reboot...

# Logs
Please attach in separate files: mirror output, library logs and kernel [logs](https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan).
In case of BSOD, please attach minidump or dump analyze output.

**Comment by Rondom** [comment]
What is your exact windows version (output of `ver` command). I am pretty sure, I have been running win-sshfs under win10 without any problems. Will double-check tomorrow

**Comment by Rondom** [comment]
Indeed it 1.0.5 is creating problems. 1.0.3 worked fine.

**Comment by KyleSanderson** [comment]
Thanks for confirming Rondom. Hopefully @Liryna is able to make another 1.0.x release as 1.1.0 crashes instantly.

**Comment by Liryna** [comment]
Hi @KyleSanderson ,

Thanks you for the report.

@Rondom , have you been able to reproduce in a VM ? Simple sleep and wake up produce the issue ?

@KyleSanderson , This will be fixed in a 1.1.x, I do not plan to handle multiple version 😢

**Comment by KyleSanderson** [comment]
@Liryna that's tragic man, any idea why there's no stable API? It looks like WinSSHFS has been battling with breaking API/ABI changes for years.

**Comment by Liryna** [comment]
@KyleSanderson API has been only break for 0.8.0 (Dec 10, 2015) and last month for 1.1.0 with a very tiny change that can be fixed in 5 sec by updating nuget package.
We know that API changes are very annoying that's why we only do it when it is really necessary.

**Comment by Rondom** [comment]
@KyleSanderson [It is explicitly recommended to ship dokan1.dll](https://github.com/dokan-dev/dokany/wiki/How-to-package-your-application-with-Dokan), which is what win-sshfs should do. This way it can continue functioning as long as the Dokan 1.x.x driver is installed.

**Comment by Liryna** [comment]
Hi @KyleSanderson ,

This should be tested with 1.1.0. You can do it without updating winsshfs:
- Copy 1.0.5/1.0.3 dokan1.dll in winsshfs folder.
- Remove dokan 1.0.5/1.0.3.
- Install dokan 1.1.0.
- Run winsshfs.

Dokan 1.1.0 has a deadlock fix that has been introduced from 1.0.3 to 1.0.5. There is a change that is already solving the issue.

**Comment by RustyHacksaw** [comment]
Hi all, this bug has new behaviour ... using the mirror app, if I mount a drive and then sleep/wake the computer, about 50% of the time, dokany will exit with an error ... debug output from the mirror app shows  

Ioctl failed for wait with code 995.
Ioctl failed for wait with code 995.
Ioctl failed for wait with code 995.
Ioctl failed for wait with code 995.
Ioctl failed for wait with code 995.
Thread will be terminated.
Thread will be terminated.
Thread will be terminated.
Thread will be terminated.
Thread will be terminated.
Dokan Error: DokanKeepAlive CreateFile failed \\.\Volume{ < GUID > }: 2
Unmounted

unload
Success

**Comment by marinkobabic** [comment]
Driver Log is needed.

**Comment by RustyHacksaw** [comment]
Yup, working on it ... windows 10 with secure boot on is a huge PITA.

**Comment by RustyHacksaw** [comment]
Finally got a log ... 
[drivercrash.LOG](https://github.com/dokan-dev/dokany/files/1772835/drivercrash.LOG)

SecureBoot is disabled and this was on Windows 10 build 16299.248

at around line 32219 it logs that it seems to be awaken from sleep mode and that's where stuff seems to go wrong and it does an unmount on it's own.

**Comment by RustyHacksaw** [comment]
Hi all, I'm trying to understand this bug ... the DokanTimeoutThread has a time check with a pretty arbitrary value of ( DOKAN_CHECK_INTERVAL + 2000 ) * 10000  - which is 70 seconds I think and if the current time and last time are greater than that, it thinks that it has woken up from a sleep event.  Anything less than that and it assumes it's a regular event, does the ReleaseTimeoutPendingIrp and then DokenCheckKeepAlive which unmounts the mounted drive.  I'm not a driver expert, but it seems like there should be a way to catch/use the IRP_MJ_POWER notification when the system goes to sleep or wakes up, but I'm not sure how to integrate it into the timeout thread code.  If the IRP_MJ_POWER is added to the main dokan dispatch routine, can we set a flag in the DCB that can be read by the DokanTimeoutThread function to have it continue to wait instead of unmounting??  Does that make sense?  Or another method might be to use KeQuerySystemTime and KeQueryUnbiasedInterruptTime to see if there is a difference?  The unbiased version is the raw time, while the normal version adds a 'bias' amount to compensate for being asleep.  Would that possibly be accurate enough?

**Comment by marinkobabic** [comment]
The main idea here was to discover that we are facing a awaken mode. As long as I know the IRP_MJ_POWER is only called when the driver supports PNP. Maybe all we need to do is here https://github.com/dokan-dev/dokany/blob/83be68351c70d916e58bdae8e3424e521fe07b04/sys/timeout.c#L294 to reset the timeout. Normally the timeout is reset using a call from user mode https://github.com/dokan-dev/dokany/blob/83be68351c70d916e58bdae8e3424e521fe07b04/sys/device.c#L990 It seems just to be a timing issue here. So resetting the timeout after the line above should fix the issue.

**Comment by RustyHacksaw** [comment]
Okay, I'll give that a try in my test env and see how it goes, thanks!

**Comment by RustyHacksaw** [comment]
Hmm, nope, it still happens.

**Comment by RustyHacksaw** [comment]
Okay, so the new log shows it's not even hitting that "awaken from sleep..." section of code, hence why it's not resetting the TickCount...  I'll dump the tickcounts now.

**Comment by RustyHacksaw** [comment]
Okay, I think it's fixed, I added the  DokanUpdateTimeout function inside that else statement so that it executes for both cases and when I test it, I can no longer reproduce the unmount.

[https://github.com/dokan-dev/dokany/blob/83be68351c70d916e58bdae8e3424e521fe07b04/sys/timeout.c#L285-L300](url)

the DokanUpdateTimeout call is inserted between lines 299 and 300.  I will create a pull request soon.

**Comment by marinkobabic** [comment]
If I understand your idea correct then the drive will never disconnected
even if the User Mode application is not running anymore.

You can test the behaviour when killing your app using task manager.

**Comment by RustyHacksaw** [comment]
Hmmm, okay, I'll test that too.

**Comment by RustyHacksaw** [comment]
Dang, yup, it never unmounts.  Sigh, so I'll move it to just inside the 'awake...' comment and see if that still works ... pretty sure that's where I had it the first time and it didn't work ... How does the User Mode signal the driver to keep the tick count updating?

**Comment by marinkobabic** [comment]
There is a thread sending a special request to the driver. Here is then the
entry point in the driver
https://github.com/dokan-dev/dokany/blob/83be68351c70d916e58bdae8e3424e521fe07b04/sys/device.c#L990

**Comment by RustyHacksaw** [comment]
Yeah, the time check in the timeout.c isn't getting hit every time, only some of the time, so resetting the timeout in only that spot doesn't work, we need a more reliable way to detect if a sleep/wake action has occurred.  Seems unique to the most recent builds of Windows 16299 as well, so microsoft has clearly made more dumb changes that is breaking this.

**Comment by KyleSanderson** [comment]
Hey all, first and foremost thank you all for working on this tragic regression on the Dokany project. 

The last "good" version that I can tell is 1.0.3 which properly unmounts the volumes on sleep. I don't think this is a new issue, rather a regression with a cset in the tree; if that helps at all.

**Comment by marinkobabic** [comment]
Would be helpful to get such a device for few weeks. I have no device where I can reproduce the issue.

**Comment by ishanarora** [comment]
> Yeah, the time check in the timeout.c isn't getting hit every time, only some of the time, so resetting the timeout in only that spot doesn't work, we need a more reliable way to detect if a sleep/wake action has occurred.

Instead of relying on the time check ( in the timeout.c ) for ignoring sleep states, #753 fixes this by ignoring one-off timeout failure, instead requiring multiple consecutive timeouts.

**Comment by Liryna** [comment]
I just import changes that remove Dokan keep alive thread and replaced to a file keepalive. 
https://github.com/dokan-dev/dokany/commit/1fdf9fb23e20d75659bf8344ce8d19f1bbcda7c8

These changes will totally remove the reason why the FS is unmount currently after a sleep.
Does anyone could try to build the current master and give a feedback that it really fix on their side ?

**Comment by Liryna** [comment]
1.2.1 has been released 🏆 https://github.com/dokan-dev/dokany/releases/tag/v1.2.1.1000 with the new keepalive that should fix this behavior.

**Comment by ishanarora** [comment]
Hi @Liryna ,
1fdf9fb looks good to me. I have not tested it though.
